notebooks/ch11/gradient_stochastic.ipynb

```diff
@@ -0,0 +1,285 @@
+    "\n",
+    "\n",
+    "### Selecting the Mini-Batch Size\n",
+    "\n",
+    "Mini-batch gradient descent is most optimal when running on a Graphical Processing Unit (GPU) or on distributed systems. Since computations on these hardware machines can be executed in parallel, using a mini-batch can increase the accuracy of the gradient without increasing computation time. Depending on the memory of the GPU, the mini-batch size is often set between 10 and 100 observations. \n",
+    "\n",
+    "### Defining a Function for Mini-Batch Gradient Descent\n",
+    "\n",
+    "A function for mini-batch gradient descent requires the ability to select a batch size. Below is a function that implements this feature."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2018-06-26T05:26:51.291775Z",
+     "start_time": "2018-06-26T05:26:51.284067Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def minimize_mini_batch(cost_fn, grad_cost_fn, dataset, minibatch_size, alpha=0.2, progress=True):\n",
+    "    \"\"\"\n",
+    "    Uses mini-batch gradient descent to minimize cost_fn.\n",
+    "    Returns the minimizing value of theta once theta changes\n",
+    "    less than 0.001 between iterations.\n",
+    "    \"\"\"\n",
+    "    NUM_OBS = len(dataset)\n",
+    "    assert minibatch_size < NUM_OBS\n",
+    "    \n",
+    "    theta = 0\n",
+    "    np.random.shuffle(dataset)\n",
+    "    while True:\n",
+    "        for i in range(0, NUM_OBS, minibatch_size):\n",
+    "            if progress:\n",
+    "                print(f'theta: {theta:.2f} | cost: {cost_fn(theta, dataset):.2f}')\n",
+    "            \n",
+    "            mini_batch = dataset[i:i+minibatch_size]\n",
+    "            gradient = grad_cost_fn(theta, mini_batch)\n",
+    "            new_theta = theta - alpha * gradient\n",
+    "            \n",
+    "            if abs(new_theta - theta) < 0.1:\n",
+    "                return new_theta\n",
+    "            \n",
+    "            theta = new_theta\n",
+    "        np.random.shuffle(dataset)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Summary\n",
+    "\n",
+    "We use batch gradient descent to iteratively improve model parameters until the model achieves low cost. Since batch gradient descent is computationally intractable with large datasets, we often use stochastic gradient descent to fit models instead. With a GPU, mini-batch gradient descent can converge more quickly than stochastic gradient descent.\n",
+    "\n",
+    "Although stochastic gradient descent typically takes more individual updates to converge than batch gradient descent, stochastic gradient descent often converges more quickly because it computes updates faster."
```

Comment:
This last line "Although stochastic gradient descent ..." addresses my comment above about clarifying that SGD may take more individual steps but still converge faster. I think this sentence would be better placed after the two images, and you could just have a short line in the summary that SGD is computationally faster than batch GD.

Reply:
👍